src/sender_queue/mod.rs

```diff
@@ -0,0 +1,366 @@
+//! # Sender queue
+//!
+//! A sender queue allows a `DistAlgorithm` that outputs `Epoched` messages to buffer those outgoing
+//! messages based on their epochs. A message is sent to its recipient only when the recipient's
+//! epoch matches the epoch of the message. Thus no queueing is required for incoming messages since
+//! any incoming messages with non-matching epochs can be safely discarded.
+
+mod message;
+
+use std::cmp::Ordering;
+use std::collections::{btree_map::Entry, BTreeMap, BTreeSet};
+use std::fmt::Debug;
+
+use rand::Rand;
+use serde::{de::DeserializeOwned, Serialize};
+
+use {DistAlgorithm, Epoched, NodeIdT, Target, TargetedMessage};
+
+pub use self::message::{Message, MessageContent};
+
+/// Generic sender queue functionality.
+pub trait SenderQueueFunc<D>
+where
+    D: DistAlgorithm,
+    <D as DistAlgorithm>::Message: Clone + Epoched + Serialize + DeserializeOwned,
+    D::NodeId: NodeIdT + Rand,
+{
+    type Step;
+
+    /// Returns a pair containing
+    ///
+    /// - the maximum epoch out of a given epoch and the epoch of a given batch,
+    ///
+    /// - an optional new peer to be added to the set of peers.
+    fn max_epoch_with_batch(
+        &self,
+        epoch: <D::Message as Epoched>::Epoch,
+        batch: &D::Output,
+    ) -> (<D::Message as Epoched>::Epoch, Option<D::NodeId>);
```

Comment:
This method seems to do two things at once. Maybe it would be cleaner to only return an `Option<D::NodeId>`, and to implement `Epoched` for the `Step`s? So instead of

```rust
let (max_e, added) = alg.max_epoch_with_batch(epoch, batch);
```

the caller would write:

```rust
let max_e = epoch.max(batch.epoch());
let added = alg.added_node(&batch);
```

Reply:
Having `Batch::next_epoch` would help because in fact we need
```
let max_e = epoch.max(batch.next_epoch);
```
It's not the epoch of the batch we are after but its next epoch.

Reply:
An issue with making `D::Output: Epoched` is that `epoch` is not of type `<D::Output as Epoched>::Epoch` but of type `<D::Message as Epoched>::Epoch`! And those are not equal.